### PR TITLE
refactor: use traditional parameters where possible

### PR DESCRIPTION
**Pull request by imnotjames**
### Description of change

this refactors a wide variety of use-cases for parameters.

* drops the native parameters out of use but leaves them in place for compatibility purposes
* refactors the parameter replacement to improve performance drastically
* allows for nested querybuilders without parameter overwriting (such as with `Brackets`
* aligns the escape parameters functions across drivers

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [ ] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/query-builder/QueryBuilder.ts [hunk 1]
```diff
@@ -72,6 +72,8 @@ export abstract class QueryBuilder<Entity> {
      */
     protected parentQueryBuilder: QueryBuilder<any>;
 
+    private parameterIndex = 0;
```
Comment (imnotjames) [inline src/query-builder/QueryBuilder.ts:75]: add comment explaining what this is

**Comment by imnotjames** [comment]
I want to do some back of the napkin benchmarking before merging this in..

**Comment by pleerock** [comment]
> but leaves them in place for compatibility purposes

can you please describe what compatibility purposes left?

**Comment by imnotjames** [comment]
> > but leaves them in place for compatibility purposes
> 
> can you please describe what compatibility purposes left?

It's a public API and could be used by third parties.

**Comment by imnotjames** [comment]
> I want to do some back of the napkin benchmarking before merging this in..

Back of the napkin, no negligible difference.  Timing was incredibly hard to see a difference over 10m inserts in batches of 100 & 10,000

**Comment by imnotjames** [comment]
PS without the changes to the way we're doing replacement escapes, the difference was that for batches of 100 ~200x slower, 1,000 it never finished, 10,000 it crashed because of too many matchers.

**Pull request merged**